glusterd2/commands/snapshot/snapshot-create.go

```diff
@@ -378,53 +381,106 @@ func validateSnapCreate(c transaction.TxnCtx) error {
 
 		return errors.New("one or more brick is not compatable")
 	}
-	if nodeData, err = populateBrickMountData(volinfo, req.SnapName); err != nil {
+	if nodeData, err = populateSnapBrickMountData(volinfo, req.SnapName); err != nil {
+
 		return err
 	}
 	c.SetNodeResult(gdctx.MyUUID, snapshot.NodeDataTxnKey, &nodeData)
 	//Quorum check ?
 	return nil
 }
+func takeVolumeSnapshots(newVol, oldVol *volume.Volinfo) error {
+	var wg sync.WaitGroup
+	numBricks := len(oldVol.GetBricks())
+	errCh := make(chan error, numBricks)
+	for subvolCount, subvol := range oldVol.Subvols {
+		for count, b := range subvol.Bricks {
+			if !uuid.Equal(b.PeerID, gdctx.MyUUID) {
+				continue
+			}
+			wg.Add(1)
+			snapBrick := newVol.Subvols[subvolCount].Bricks[count]
+			go brickSnapshot(errCh, &wg, snapBrick, b)
+
+		}
+	}
+	go func() {
+		wg.Wait()
```

Comment:
Waiting on goroutines by launching another goroutine does not seem like a good idea. You should wait in a goroutine that you have control over.

I suggest that you collect the errors in a different goroutine (the loop below) and wait for the snapshot goroutines and the err collect goroutine in this function.